### PR TITLE
CAT-2371 Correction in Textures Disposal

### PR DESCRIPTION
**Pull request by nishantt95**
Fixing a minor sub-bug to make correction to dispose of the texture

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java [hunk 1]
```diff
@@ -811,20 +811,25 @@ private void initScreenMode() {
...
-			for (int j = 0; j < dataSize; j++) {
-				LookData lookData = data.get(j);
-				lookData.getPixmap().dispose();
-				lookData.getTextureRegion().getTexture().dispose();
+	private void disposeScenes() {
+		for (Scene scene : ProjectManager.getInstance().getCurrentProject().getSceneList()) {
+			List<Sprite> sprites = scene.getSpriteList();
+			for (Sprite sprite : sprites) {
+				List<LookData> data = sprite.getLookDataList();
+				disposeTextures(data);
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java]: * `disposeScenes` isn't a good name. Overload `disposeTextures` to take a Project object instead.
* Inline `scene.getSpriteList()`
* `ProjectManager.getInstance().getCurrentProject()` => `project`

File: catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java [hunk 2]
```diff
@@ -812,19 +812,23 @@ private void initScreenMode() {
...
 
+	private void disposeTextures(List<LookData> data) {
+		int dataSize = data.size();
+		for (int j = 0; j < dataSize; j++) {
+			LookData lookData = data.get(j);
+			lookData.getPixmap().dispose();
+			lookData.getTextureRegion().getTexture().dispose();
+		}
+	}
+
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java]: As mentioned in the Email, please only use a single disposeTextures method. There is no benefit in splitting it up.

File: catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java [hunk 3]
```diff
@@ -812,15 +812,16 @@ private void initScreenMode() {
...
+		for (Scene scene : project.getSceneList()) {
+			List<Sprite> sprites = scene.getSpriteList();
+			for (Sprite sprite : sprites) {
+				List<LookData> lookDatas = sprite.getLookDataList();
+				int lookDatasSize = lookDatas.size();
+				for (int j = 0; j < lookDatasSize; j++) {
+					LookData lookData = lookDatas.get(j);
+					lookData.getPixmap().dispose();
+					lookData.getTextureRegion().getTexture().dispose();
+				}
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java]: Improve the lookDatas loop by using for-each.
Inline all variables. There is no need to declare `List<?>`

File: catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java [hunk 4]
```diff
@@ -812,15 +823,13 @@ private void initScreenMode() {
...
-		for (int i = 0; i > spriteSize; i++) {
-			List<LookData> data = sprites.get(i).getLookDataList();
-			int dataSize = data.size();
-			for (int j = 0; j < dataSize; j++) {
-				LookData lookData = data.get(j);
-				lookData.getPixmap().dispose();
-				lookData.getTextureRegion().getTexture().dispose();
+		for (Scene scene : project.getSceneList()) {
+			for (Sprite sprite : scene.getSpriteList()) {
+				for (int j = 0; j < sprite.getLookDataList().size(); j++) {
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java]: `for (LookData lookData : sprite.getLookDataList()) {`

File: catroid/src/androidTest/java/org/catrobat/catroid/test/stage/StageListenerTest.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		when(lookData.getTextureRegion()).thenReturn(textureRegion);
+		when(textureRegion.getTexture()).thenReturn(texture);
+
+		StageListener stageListener = new StageListener(project);
+		stageListener.textureDisposal();
+
+		verify(lookData.getPixmap(), times(1)).dispose();
+		verify(lookData.getTextureRegion().getTexture(), times(1)).dispose();
+	}
+}
```
Comment (robertpainsi) [inline catroid/src/androidTest/java/org/catrobat/catroid/test/stage/StageListenerTest.java]: Please remove the test. I tried to test it cleanly but the StageListener in it's current state isn't unit testable.

JFI: you don't need to mock everything :laughing:
If it's easy and fast to create an object, don't mock it. If you have to verify behavior which can't be checked by public interfaces, mock it.
So in the end you just mock `LookData` and verify that `LookData.resetLookData()` is called. You don't need to retest that the resources are freed as it's already tested in `LookDataTest.testPixmapAndTextureRegionDisposal`

Comment (nishantt95) [inline catroid/src/androidTest/java/org/catrobat/catroid/test/stage/StageListenerTest.java]: > JFI: ...

Thanks 

So, do I have to keep the test or not?

Comment (robertpainsi) [inline catroid/src/androidTest/java/org/catrobat/catroid/test/stage/StageListenerTest.java]: Please remove the test

File: catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java [hunk 6]
```diff
@@ -26,6 +26,7 @@
 import android.graphics.Bitmap.Config;
 import android.graphics.PointF;
 import android.os.SystemClock;
+import android.support.annotation.VisibleForTesting;
```
Comment (robertpainsi) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageListener.java]: Please remove any unnecessary change

**Comment by robertpainsi** [comment]
test this please

**Comment by robertpainsi** [comment]
please run espresso tests

**Comment by robertpainsi** [comment]
test this please

**Comment by robertpainsi** [comment]
please run espresso tests
